frame/staking/src/pallet/impls.rs

```diff
@@ -879,21 +887,70 @@ impl<T: Config> ElectionDataProvider for Pallet<T> {
 	type AccountId = T::AccountId;
 	type BlockNumber = BlockNumberFor<T>;
 	type MaxVotesPerVoter = T::MaxNominations;
+	type Pages = T::Pages;
 
 	fn desired_targets() -> data_provider::Result<u32> {
 		Self::register_weight(T::DbWeight::get().reads(1));
 		Ok(Self::validator_count())
 	}
 
-	fn electing_voters(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<VoterOf<Self>>> {
-		// This can never fail -- if `maybe_max_len` is `Some(_)` we handle it.
-		let voters = Self::get_npos_voters(maybe_max_len);
-		debug_assert!(maybe_max_len.map_or(true, |max| voters.len() <= max));
+	/// implementation notes: Given the loose dynamics of the `ElectionProvider`, namely the fact
+	/// that number of entries returned per page is dynamic, this implementation is best-effort. We
+	/// try and use the `LastIteratedNominator`, which is always kept as long call with `remaining >
+	/// 0` are made. Any call with `remaining = 0` will clear this history item, and the next call
+	/// will be a fresh start.
+	fn electing_voters_paged(
+		maybe_max_len: Option<usize>,
+		mut remaining: PageIndex,
+	) -> data_provider::Result<Vec<VoterOf<Self>>> {
+		remaining = remaining.min(Self::msp());
+
+		let mut maybe_last = LastIteratedNominator::<T>::get();
+		let is_valid_state = if T::Pages::get() > 1 {
+			// in the first page, the `LastIteratedNominator` should not exist, in the rest of the
+			// pages it should.
+			(remaining == Self::msp()) ^ maybe_last.is_some()
```

Comment:
Interesting, so this bitwise XOR basically is also a logical XOR? I assume its equivalent to (!A && B) || (A && !B)